### PR TITLE
CDAP-14810 Add scan using index API to storage SPI

### PR DESCRIPTION
**Pull request by poornachandra**
JIRA - https://issues.cask.co/browse/CDAP-14810
Build - https://builds.cask.co/browse/CDAP-DUT6775

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTable.java [hunk 1]
```diff
@@ -178,16 +190,10 @@ public void close() throws IOException {
    * Convert the columns to corresponding byte array, each column has to be part of the schema.
    *
    * @param columns columns to convert
-   * @return the converted byte array, null if read for all columns
+   * @return the converted byte array
    * @throws InvalidFieldException some column is not part of the schema
    */
-  @Nullable
-  private byte[][] convertColumnsToBytes(@Nullable Collection<String> columns) throws InvalidFieldException {
-    // Empty columns means to read all columns. The corresponding parameter for Table is null
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTable.java:186]: is this an unrelated fix? 
I am not sure I understand why it is not nullable?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableAdmin.java [hunk 2]
```diff
@@ -75,14 +81,14 @@ public void drop(StructuredTableId tableId) throws IOException {
     LOG.info("Dropping table {} in namespace {}", tableId.getName(), NamespaceId.SYSTEM);
     StructuredTableSpecificationRegistry.removeSpecification(tableId);
     if (StructuredTableSpecificationRegistry.isEmpty()) {
-      DatasetAdmin admin = tableDefinition.getAdmin(SYSTEM_CONTEXT, entityTableSpec, null);
-      LOG.info("Dropping dataset table {} in namespace {}", entityTableSpec.getName(), NamespaceId.SYSTEM);
+      DatasetAdmin admin = indexTableDefinition.getAdmin(SYSTEM_CONTEXT, indexTableSpec, null);
+      LOG.info("Dropping dataset table {} in namespace {}", indexTableSpec.getName(), NamespaceId.SYSTEM);
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableAdmin.java]: wanna say, indexed table?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableAdmin.java [hunk 3]
```diff
@@ -45,22 +47,26 @@
...
   public void create(StructuredTableSpecification spec) throws IOException {
-    LOG.info("Creating table {} in namespace {}", spec.getTableId().getName(), NamespaceId.SYSTEM);
-    DatasetAdmin admin = tableDefinition.getAdmin(SYSTEM_CONTEXT, entityTableSpec, null);
-    if (!admin.exists()) {
-      LOG.info("Creating dataset table {} in namespace {}", entityTableSpec.getName(), NamespaceId.SYSTEM);
-      admin.create();
+    LOG.info("Creating table {} in namespace {}", spec, NamespaceId.SYSTEM);
+    DatasetAdmin indexTableAdmin = indexTableDefinition.getAdmin(SYSTEM_CONTEXT, indexTableSpec, null);
+    if (!indexTableAdmin.exists()) {
+      LOG.info("Creating dataset table {} in namespace {}", indexTableSpec.getName(), NamespaceId.SYSTEM);
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableAdmin.java]: wanna say, indexed table?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/PostgresSqlStructuredTable.java [hunk 4]
```diff
@@ -129,6 +129,28 @@ public void upsert(Collection<Field<?>> fields) throws InvalidFieldException, IO
...
 
+  @Override
+  public CloseableIterator<StructuredRow> scan(Field<?> index) throws InvalidFieldException, IOException {
+    LOG.trace("Table {}: Scan index {}", tableSchema.getTableId(), index);
+    fieldValidator.validateField(index);
+    if (!tableSchema.isIndexColumn(index.getName())) {
+      throw new InvalidFieldException(tableSchema.getTableId(), index.getName(), "is not an indexed column");
+    }
+
+    String sql = getReadQuery(Collections.singleton(index), Collections.emptySet(), false);
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/PostgresSqlStructuredTable.java]: I feel like it would be cleaner to pass in null for the columns to read. 
emptySet() is like saying, I want no columns except for the primary keys.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/PostgresSqlStructuredTableAdmin.java [hunk 5]
```diff
@@ -70,10 +84,12 @@ public StructuredTableSpecification getSpecification(StructuredTableId tableId)
 
   @Override
   public void drop(StructuredTableId tableId) throws IOException {
+    LOG.info("Creating table {}", tableId);
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/PostgresSqlStructuredTableAdmin.java]: Dropping table?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/PostgresSqlStructuredTableAdmin.java [hunk 6]
```diff
@@ -46,15 +51,24 @@ public PostgresSqlStructuredTableAdmin(DataSource dataSource) {
 
   @Override
   public void create(StructuredTableSpecification spec) throws IOException, TableAlreadyExistsException {
+    LOG.info("Creating table {}", spec);
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/PostgresSqlStructuredTableAdmin.java]: I think you should only log this if you actually create it, that is, if it does not already exist,

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/SqlTransactionRunner.java [hunk 7]
```diff
@@ -60,7 +60,9 @@ public void run(TxRunnable runnable) throws TransactionException {
         connection.rollback();
         throw new TransactionException("Failed to execute the sql queries.", e);
       } catch (SQLException sql) {
-        throw new TransactionException("Failed to rollback a transaction.", sql);
+        TransactionException ex = new TransactionException("Failed to rollback a transaction.", sql);
+        ex.addSuppressed(e);
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/SqlTransactionRunner.java]: seems like it should be the other way round: e.addSuppressed(ex). Because e is the original exception, and ex is or may be misleading as the cause of error.

File: cdap-storage-spi/src/main/java/co/cask/cdap/spi/data/table/StructuredTableSpecification.java [hunk 8]
```diff
@@ -221,16 +223,23 @@ private void validate() throws InvalidFieldException {
           throw new InvalidFieldException(tableId, primaryKey);
         }
         if (!Fields.isPrimaryKeyType(type)) {
-          throw new InvalidFieldException(tableId, primaryKey, "is not defined as primary key column");
+          throw new InvalidFieldException(
+            tableId, primaryKey,
+            String.format("has wrong type for a primary key. Valid types are: %s", FieldType.PRIMARY_KEY_TYPES));
```
Comment (anew) [inline cdap-storage-spi/src/main/java/co/cask/cdap/spi/data/table/StructuredTableSpecification.java:228]: +1 for better error messages

**Review by anew — COMMENTED** [review]
Please see my mostly minor comments

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTable.java (hunk 1 above)
Comment (poornachandra) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTable.java:186]: In a previous PR, I had refactored the read method for all columns to not use this method. So null would never get passed to this method anymore.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableAdmin.java (hunk 3 above)
Comment (poornachandra) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableAdmin.java]: +1

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableAdmin.java (hunk 2 above)
Comment (poornachandra) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableAdmin.java]: +1

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/PostgresSqlStructuredTable.java (hunk 4 above)
Comment (poornachandra) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/PostgresSqlStructuredTable.java]: +1

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/PostgresSqlStructuredTableAdmin.java (hunk 6 above)
Comment (poornachandra) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/PostgresSqlStructuredTableAdmin.java]: +1

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/PostgresSqlStructuredTableAdmin.java (hunk 5 above)
Comment (poornachandra) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/PostgresSqlStructuredTableAdmin.java]: +1

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/SqlTransactionRunner.java (hunk 7 above)
Comment (poornachandra) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/SqlTransactionRunner.java]: +1

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTable.java (hunk 1 above)
Comment (anew) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTable.java:186]: ok got it

**Review by anew — APPROVED** [review]

**Comment by poornachandra** [comment]
Build is green
